libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectEx.kt

```diff
@@ -0,0 +1,65 @@
+@file: JvmName("RoadObjectEx")
+
+package com.mapbox.navigation.core.trip.roadobject
+
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.base.trip.model.eh.EHorizon
+import com.mapbox.navigation.base.trip.model.roadobject.UpcomingRoadObject
+import com.mapbox.navigation.core.trip.roadobject.reststop.RestStopFactory
+import com.mapbox.navigation.core.trip.roadobject.reststop.RestStopFromIntersection
+import com.mapbox.navigation.core.trip.roadobject.reststop.RestStopFromRoadObject
+import com.mapbox.navigation.utils.internal.ifNonNull
+
+/**
+ * Use the API to retrieve a list of all upcoming rest stops using [UpcomingRoadObject].
+ * This is recommended to be used with:
+ * - [EHorizon] during Free Drive or
+ * - [RouteProgress] during Active Navigation.
+ */
+fun List<UpcomingRoadObject>.getAllRestStops(): List<RestStopFromRoadObject> {
+    return RestStopFactory.createWithUpcomingRoadObjects(this)
+}
+
+/**
+ * Use the API to retrieve the first upcoming rest stops using [UpcomingRoadObject].
+ * This is recommended to be used with:
+ * - [EHorizon] during Free Drive or
+ * - [RouteProgress] during Active Navigation.
+ */
+fun List<UpcomingRoadObject>.getFirstUpcomingRoadObject(): RestStopFromRoadObject? {
+    val restStops = this.getAllRestStops()
+    if (restStops.isNotEmpty()) {
+        val distanceToStart = restStops.first().distanceToStart
+        if (distanceToStart != null && distanceToStart > 0) {
+            return restStops.first()
+        }
+    }
+    return null
+}
+
+/**
+ * Use the API to retrieve the list of all rest stops using [NavigationRoute].
+ * This is recommended to be used with:
+ * - [NavigationRoute] during Route Preview
+ */
+fun NavigationRoute.getAllRestStops(): Map<RouteLeg, List<RestStopFromIntersection>> {
```

Comment:
it would be hard to exted later if needed, what if we add wrapper like
```kotlin
class RouteRestsStops() {
  /**route id*/
  routeId: Int
  /**index of leg and rest stops*/
  val legsRestStops: Map<Int, List<RestStopFromIntersection>
}
```

Also, can be extended for Incidents and other things 
` this PR will serve as the base for exposing other road objects down the line`
```
class RouteExtra<T> -> class RouteExtra<RestStopFromIntersection>
class RouteExtra<T> -> class RouteExtra<incidents>
```